### ISSUE TITLE
能否公开下您构建 CheXpert HDF5 的数据处理脚本？

### ISSUE THREAD

**Issue by Hessiuue** [YuzunoKawori/MCD-CLIP#1]
您好，感谢您开源这项非常棒的工作！
我最近在复现论文 《Multi-view Classification Using Hybrid Fusion and Mutual Distillation》 的实验时，遇到一些结果对不上的情况，想请教您关于 CheXpert 数据处理和训练细节 的部分问题。

一、环境信息
PyTorch：2.8.0+cu128
CUDA：12.8
GPU：RTX 4090
系统：Windows 11

训练脚本：官方仓库中的 train.py（未改动核心逻辑）

数据集：CheXpert v1.0（筛选双视角后约 31,413 studies）

二、我目前的数据处理流程（自己写的 HDF5 转换脚本）

我按照论文描述和其他实现，自己写了一个 build_chexpert_h5.py 脚本，把 CheXpert 转成 HDF5 格式。
主要做法如下：

双视角筛选：只保留同一 study 下同时有 frontal 和 lateral 的样本；如果有多张同类视角图像，就取编号最小的 viewN_*。

图像处理：转为灰度图，resize 到 224×224。

不确定标签五种场景：

U-Ignore / U-Zeros / U-Ones / U-SelfTrained / U-MultiClass；

unknown 记为 99；

U-MultiClass 用 0/1/2 表示 neg/pos/unc；

U-SelfTrained 保留 soft 标签占位，后续可回填教师概率。

划分：按患者划分 75 / 12.5 / 12.5（避免患者泄漏）。

输出文件结构：chexpert_train/val/test.h5，路径为 scans/<patient>/<study>/，包含 view1、view2 和对应标签。

处理完后，我这边得到双视角样本约 31k，对应大约 22k 患者。
脚本我已经附在 issue 里，您可以直接参考。

三、复现实验结果（Atelectasis）

我分别在五个不确定性场景下独立训练并测试，结果如下：

场景	                Test AUC	Accuracy
U-Ignore   	0.6192	91.39%
U-Zeros    	0.5455	53.82%
U-Ones	        0.6179	95.59%
U-SelfTrained	0.6674	92.37%
U-MultiClass	0.6988	92.37%
  平均 AUC	0.62976	—

论文中对应的平均 AUC 是 0.759，差距大约 0.13。
从趋势上看顺序一致（U-MultiClass > U-SelfTrained > U-Ignore ≈ U-Ones > U-Zeros），但整体偏低。

四、训练日志（Atelectasis / U-Zeros）
the number of train set: 5691
the number of validation set: 953
the number of test set: 929

===== Epoch 40/120 Summary =====
Train Loss: 0.7084
Val  Loss: 0.6976
Val  AUC : 0.5411
Best AUC : 0.5536

===== Epoch 80/120 Summary =====
Train Loss: 0.7112
Val  Loss: 0.6854
Val  AUC : 0.5480
Best AUC : 0.5536

===== Epoch 120/120 Summary =====
Train Loss: 0.6974
Val  Loss: 0.7025
Val  AUC : 0.5540
Best AUC : 0.5540

Test AUC: 0.5455
Accuracy: 53.82 %

模型训练得很平稳，但 AUC 基本不涨。感觉可能是没收敛完全，或者学习率 / epoch 数不太匹配。

五、想请教的问题 

能否方便公开一下您当时生成 CheXpert HDF5 数据的脚本？
就算只给一个简化版或关键片段也非常有帮助。
（主要是想确认视角筛选、标签映射、归一化方式和文件结构是不是和论文一致）

如果不能公开完整脚本，能否说明下 HDF5 的结构？比如：
<file>.h5
└─ scans/
   └─ patientXXXX/
      └─ studyY/
         ├─ view1
         ├─ view2
         ├─ <Task>-U-Ignore
         ├─ <Task>-U-Zeros
         ├─ <Task>-U-Ones
         ├─ <Task>-U-SelfTrained
         └─ <Task>-U-MultiClass

希望能确认数据类型（uint8 / float32）、是否做过归一化、是否压缩（gzip / chunking），以及任务列命名是否与 CSV 完全一致。
论文中训练 120 epoch 是最终设定吗？我看日志中 AUC 一直没明显提升，是否需要更长训练？
您是否用了 warmup、Cosine LR 调度、或 focal loss / 类别权重？
输入分辨率是 224×224 还是更高（384 或 390）？

再次感谢您提供代码和论文！
我非常希望能对齐您当时的数据处理方式，因为这部分对最终性能影响挺大。
如果您能分享一下 HDF5 构建脚本或结构说明，将会非常有帮助 

祝好，
Hessiuue

**Comment by YuzunoKawori** [comment]
很高兴您对我们的工作感兴趣并花时间去复现！ 

1.关于CheXpert HDF5文件的生成，我们大体按照 MICCAI 2021 论文：Cross-view transformers for multi-view analysis of unregistered medical images作者提供的官方库来构建，您可以参考https://github.com/gvtulder/cross-view-transformers。 值得注意的是原文的分辨率是390×390，而我们的方法的分辨率是**224×224**。在这里我们采用了两种方法去尝试：1. 继续采用裁剪的方式将分辨率修改为224×224，好处是能够屏蔽到大部分背景区域，缺点是可能会裁剪掉有效区域。 2.我们使用resize的方式修改分辨率，这样能够更好地保留有效区域但同时会损失一些细粒度特征。经过实验这两种处理方法性能差距不大，第一种效果会好一点，您可以根据需要来灵活使用。我们也使用了平均值方差的归一化方法，并没有进行额外的压缩。

2.针对实验结果，论文中我们提到其实我们只训练了60 epochs，这其实是出于高效考虑，在实际实验中我们观察到一些task在60 epochs之后性能还会有一定的提升，故我们在上传的代码中设置成了120 epochs（但这实际上是冗余的），您可以通过早停的方式来提高训练速度。对于AUC提升问题，我注意到您提供的示例是U-zeros情况下Atelectasis任务的情况，通过查看先前的运行日志，我们观察到了同样的情况，一个可能的原因是在该设置下uncertain中可能包含了更多错误标签。我们在其他的标签处理（如U-ones等）以及U-zeros下其他任务的情况下，观察到正常的AUC上升。

3.关于学习率等其他设置和论文里均保持一致，关于在其他标签处理下与我们实现性能的差距。 一个可能的原因就是数据处理和划分的不同导致了这个现象，因为正如您所说这部分对性能影响挺大，除此之外设备和运行环境的不同可能也会导致性能在小范围里浮动。

以上是针对您提出的宝贵问题的回复，希望这些能够帮助到您更好地了解我们工作！如果有任何问题欢迎随时向我们联系！ :blush: :blush: :blush:

**Comment by Hessiuue** [comment]
再次感谢您花费时间给予如此细致、耐心的回复！

您关于数据预处理（裁剪与Resize的权衡）、训练周期设置与早停法的建议，以及对于特定任务AUC表现的深入分析，都让我对您的工作有了更深刻的理解，也为我们后续的复现和实验提供了非常关键的指导。

在接下来的工作中，我们会先参考您提供的方法进行尝试。过程中如果遇到新的问题，可能还会冒昧地向您请教，希望不会打扰到您。

祝您科研顺利，期待您更多精彩的成果～

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年10月17日 14:39 |
| 收件人 | YuzunoKawori/MCD-CLIP ***@***.***> |
| 抄送人 | Hessiuue ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [YuzunoKawori/MCD-CLIP] 能否公开下您构建 CheXpert HDF5 的数据处理脚本？ (Issue #1) |
YuzunoKawori left a comment (YuzunoKawori/MCD-CLIP#1)

很高兴您对我们的工作感兴趣并花时间去复现！

1.关于CheXpert HDF5文件的生成，我们大体按照 MICCAI 2021 论文：Cross-view transformers for multi-view analysis of unregistered medical images作者提供的官方库来构建，您可以参考https://github.com/gvtulder/cross-view-transformers。 值得注意的是原文的分辨率是390×390，而我们的方法的分辨率是224×224。在这里我们采用了两种方法去尝试：1. 继续采用裁剪的方式将分辨率修改为224×224，好处是能够屏蔽到大部分背景区域，缺点是可能会裁剪掉有效区域。 2.我们使用resize的方式修改分辨率，这样能够更好地保留有效区域但同时会损失一些细粒度特征。经过实验这两种处理方法性能差距不大，第一种效果会好一点，您可以根据需要来灵活使用。我们也使用了平均值方差的归一化方法，并没有进行额外的压缩。

2.针对实验结果，论文中我们提到其实我们只训练了60 epochs，这其实是出于高效考虑，在实际实验中我们观察到一些task在60 epochs之后性能还会有一定的提升，故我们在上传的代码中设置成了120 epochs（但这实际上是冗余的），您可以通过早停的方式来提高训练速度。对于AUC提升问题，我注意到您提供的示例是U-zeros情况下Atelectasis任务的情况，通过查看先前的运行日志，我们观察到了同样的情况，一个可能的原因是在该设置下uncertain中可能包含了更多错误标签。我们在其他的标签处理（如U-ones等）以及U-zeros下其他任务的情况下，观察到正常的AUC上升。

3.关于学习率等其他设置和论文里均保持一致，关于在其他标签处理下与我们实现性能的差距。 一个可能的原因就是数据处理和划分的不同导致了这个现象，因为正如您所说这部分对性能影响挺大，除此之外设备和运行环境的不同可能也会导致性能在小范围里浮动。

以上是针对您提出的宝贵问题的回复，希望这些能够帮助到您更好地了解我们工作！如果有任何问题欢迎随时向我们联系！ 😊 😊 😊

**Comment by YuzunoKawori** [comment]
感谢您对我们工作的认可！ 在未来的时间里，如果遇到任何问题请尽管讨论。如果觉得我们的工作对您有一点点帮助的话，希望您能star让更多的人知道！

也预祝您未来科研顺利，为社区做出卓越的贡献！

**Comment by Hessiuue** [comment]
同学您好：

非常感谢您之前那么详细地回复我关于数据处理和训练细节的问题，对我帮助真的特别大！🙏

我在继续复现实验时遇到一个关于结果统计的小疑问。
我注意到论文中提到“最终结果是取五种不确定性场景（U-Ignore、U-Zeros、U-Ones、U-SelfTrained、U-MultiClass）的平均值”，但我在按这个方式计算时，整体平均 AUC 明显低于表格中的结果（下面是我的复现结果对比图）。

单场景的效果我这边是对齐的，但五个场景平均后就会拉低不少。
所以我在想，会不会论文表格里的结果其实是取了五个场景中表现最好的那个 AUC，而不是严格的平均？
当然也有可能是我这边理解或实现上有细微差别，所以想向您再确认一下。

非常感谢您一直以来的耐心指导！
希望没有打扰到您，如果方便的话，还想听听您对此的看法～

祝科研顺利，生活愉快！
Hessiuue

  

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年10月18日 01:59 |
| 收件人 | ***@***.***> |
| 抄送人 | ***@***.***>,
***@***.***> |
| 主题 | Re: [YuzunoKawori/MCD-CLIP] 能否公开下您构建 CheXpert HDF5 的数据处理脚本？ (Issue #1) |

Closed #1 as completed.

**Comment by YuzunoKawori** [comment]
您好！
很高兴您对我们的方法进行的深入的复现！ 但是我好像看不到您的复现对比图诶。如果能重新提供一下您的实验结果，将会对我们有很大的帮助。
谢谢！

**Comment by Hessiuue** [comment]
非常感谢您的回复！


我这次重新附上了对比图，展示了我复现的五场景平均结果和您文章中的对比。

我注意到一个可能的细节：在我目前的实验中，训练阶段没有开启图像增强，只是做了resize与归一化处理。
我在想，会不会是因为您在原始实验中使用了图像增强策略，从而在多场景平均时提升了模型的泛化性能？

想请您确认一下，在论文实验中是否有启用数据增强？


再次感谢您花时间查看我的复现结果，也非常感谢您之前耐心细致的解答！🙇‍♀️🙇‍♀️🙇‍♀️
祝科研顺利，生活愉快～

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年10月21日 17:19 |
| 收件人 | ***@***.***> |
| 抄送人 | ***@***.***>,
***@***.***> |
| 主题 | Re: [YuzunoKawori/MCD-CLIP] 能否公开下您构建 CheXpert HDF5 的数据处理脚本？ (Issue #1) |
YuzunoKawori left a comment (YuzunoKawori/MCD-CLIP#1)

您好！
很高兴您对我们的方法进行的深入的复现！ 但是我好像看不到您的复现对比图诶。如果能重新提供一下您的实验结果，将会对我们有很大的帮助。
谢谢！

**Comment by YuzunoKawori** [comment]
<img width="1028" height="422" alt="Image" src="https://github.com/user-attachments/assets/768550cc-e113-4ed2-a693-8eb2d57db9ba" /> 
谢谢您再一次花时间上传您的复现结果，不过应该是我这边的问题我还是看不到。如果方便的话，能否发送至我的邮箱 sonnycai4649@gmail.com？我们在论文实验的过程中采取的是默认操作，所以没有进行数据增强。
希望这个回答能帮助您！我在看完复现结果后会向您提供更多的分析。
谢谢！
